File: algorithms/0094.binary_tree_inorder_traversal/solution.ts
// Definition for a binary tree node.
class TreeNode {
  val: number
  left: TreeNode | null
  right: TreeNode | null
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val
    this.left = left === undefined ? null : left
    this.right = right === undefined ? null : right
  }
}

// 94. Binary Tree Inorder Traversal (Easy)
// https://leetcode.com/problems/binary-tree-inorder-traversal/description/

// Iterative solution
const inorderTraversal = (root: TreeNode | null): number[] => {
  const result: number[] = []
  const nodesStack: TreeNode[] = []

  let currentNode: TreeNode | null = root

  while (currentNode || nodesStack.length) {
    if (currentNode) {
      nodesStack.push(currentNode)
      currentNode = currentNode.left
    } else {
      currentNode = nodesStack.pop()

      result.push(currentNode.val)

      currentNode = currentNode.right
    }
  }

  return result
}

File: algorithms/0094.binary_tree_inorder_traversal/solution.go
package algorithms

import (
	data_structures "github.com/ocintnaf/leetcode/data_structures/stack"
)

// Definition for a binary tree node
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// 94. Binary Tree Inorder Traversal (Easy)
// https://leetcode.com/problems/binary-tree-inorder-traversal/description/

// Iterative solution
func InorderTraversal(root *TreeNode) []int {
	var result []int
	nodesStack := new(data_structures.Stack[*TreeNode])

	currentNode := root

	for currentNode != nil || !nodesStack.IsEmpty() {
		if currentNode != nil {
			nodesStack.Push(currentNode)
			currentNode = currentNode.Left
		} else {
			currentNode = nodesStack.Pop()

			result = append(result, currentNode.Val)

			currentNode = currentNode.Right
		}
	}

	return result
}

// Recursive solution
func InorderTraversalRecursive(root *TreeNode) []int {
	var result []int

	if root != nil {
		result = append(result, InorderTraversalRecursive(root.Left)...)
		result = append(result, root.Val)
		result = append(result, InorderTraversalRecursive(root.Right)...)
	}

	return result
}
